text/0024-cron.md

```diff
@@ -0,0 +1,196 @@
+* define the characteristics & requirements of running a "fog service"
+
+# Motivation
+[motivation]: #motivation
+
+As of this writing, all of my datasets are stale. They haven't been updated because, frankly, I'm too busy/distracted to remember to update them. This is a more serious issue than it seems at first. To put things bluntly: every dataset I create with Qri means more work for me, not less. This property has gotten in the way of me creating & maintaining lots of datasets. Since our goal is to convince others that there is merit in us working together on data, this makes getting to the "collaboration" part hard. 
+
+To keep datasets fresh, qri needs to automate the update process.  By _automate_ I mean: execute dataset updates as a background process, registered with the local operating system, according to some schedule I can specify. My computer should do the work of keeping datasets up to date for me, and should show me a log of what's been done for me, calling attention to significant changes.
+
+The magnitude of this shift to automation is noticeable. cron is _doing work for me_. It takes the process of dataset updating out of the categroy of chores, and opens up the possibliity that Qri can surprise me with new insights. 
+
+Adding cron starts to transition qri into territory best described as a _fog service_: a decentralized network of machines automatically producing local resources for a user that can aggregate to a broader network. Over time I'm hoping we can do more and more things under this "fog service" paradigm.
+
+Building a fog service requires special consideration to get right. This feature proposes running parts of Qri as a background process on a user's primary computing device. In this setting, we need to take great care to avoid interrupting a user's day-to-day computing experience, while also handling changing connectivity & resource availability. There is a very real chance that as a consequence of a cron job a user scheulded 6 months ago, Qri will be scheduled to start an intense script on a laptop with 5% battery left at a coffee shop while connected to restricted Wifi, while the user is nervously trying to get some dockerized python script to process a report that's due in an hour. We should aim to shape both the technical foundations and user experince of Qri to do the right thing in that context.
+
+This is a very high bar, but I think it's the standard we should work toward. Qri cron is both a big first step and a chance to iterate on meeting these requirements before including more parts of the Qri daemon in this "fog service" paradigm.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Here's a typical workflow for qri cron using datasets. This example shows working with the command line. As usual all of this would also be accessible over the local API. To start with, `qri cron` is a command with a number of subcommands. Ths first is `qri cron add`:
```

Comment:
Larger question: why keep a division between `qri cron` and `qri update`? Are they different enough that they need to be separate commands? Is it worth having to explain to users why there are two commands that do nearly the same thing? I'm reminded of our long ago discussions around `qri save` and `qri new`.

Alternatively, could `qri cron` be merged into a subset of `qri update`, like `qri update schedule me/dataset_name`? Doing so would also help solve the dependency problem, wherein cron invokes update, but update has to notify cron when run manually.

Related but separate to this question, do we want `qri cron` (whether or not it gets combined with update) to be able to fetch new data from other peers, the way that `qri update` can?

Reply:
>  do we want qri cron to be able to fetch new data from other peers, the way that qri update can?

Yes. Cron schedules and automates the qri update process, so whatever behavior you would have running `qri update` directly should be the exact same here. If `connect` is running when `udpate` is called, the d.web is available and can be used. If not, more requests fall back to the registry. This is why we really want to move toward `qrid` quickly to get easier dweb "toggling".

** ** 

>  why keep a division between qri cron and qri update?

really, really good question. @dustmop & I chatted about this a bunch offline, and agreed we see no reason to keep both. Like the save/new debate, it's confusing to have the concept of updating beside the concept of scheduling.

A few really great points came out of that discussion:

### This RFC should be a breaking change on update
`cron` isn't a very user friendly. `update` is a lot easier, so we agreed this rfc should switch to modifying the way update works instead of introducing a new term.

### the current `qri update` is a lot like the refresh button on your browser
`qri update` as it currently works is just an alias for `qri save --recall tf`, which in english is just "re-run that transform script and create a new version if something changes" . That sounds a lot like "refresh". Now that we're thinking about scheduled fetching, think about how often you hit the refresh button on your browser.

In this metaphor it's clear that scheduled updates are the default. When data is syncing to an external source, qri should be doing the sync work for you, and hand you a button that works a lot like "refresh", that lets you override qri's default for an immediate check. Dustin used the metaphor of when you've just signed up for a service, and you're refreshing your email inbox because you _know_ an email is coming.

One issue we can't really get around is "default" here only applies in situations where updates can be automated. There are lots of workflows in qri that don't work this way. This means we'll always need some sort of "automation setup" step.